Singlepage

erwindon opened the pull request:
This PR turns SaltGUI into a single-page web application.
Navigation is no longer between web-pages, but between states.
Don't worry, the states act as pages.

benefits:
* navigation is faster
* navigation has nice page transitions
* the back-button should now always work properly
* there is only one type of navigation

This may also solve one or more of #229, #293, #327.
This obsoletes #334.

erwindon commented:
@bbinet can you please give this a quick/medium/full(your choice) test?

bbinet commented:
Good job @erwindon, this is great!
The only small issue I have seen is that if I open a jobid then close it by clicking on the up-right cross, the page is correctly updated, but not the url, so if I reload the page, the jobid page is shown again.

erwindon commented:
@bbinet not that I doubted your observation, but I see the same thing. And it is consistent with the other close-buttons, e.g. when viewing the per-minion details of grains/schedules/pillars/beacons.
The handling of the close-buttons is different than other navigation because the effect needs to be exactly the same as the back-button from the browser.

erwindon commented:
And I spotted another one. adding menu items to panel-level menu's was broken. since we are now re-using panels, the menus were also re-used. but in some cased the same menu items were being added again. I chose too quickly to erase all these menu's when navigating. the 3 pages (jobs/keys/minions) that added their menu item only once in the constructor lost them very quickly. I now need to decide between the 2 solutions instead a part of each...

erwindon commented:
fixed the panel-level dropdown menu's, these now always assume the world has changed and will determine again-and-again the visibility of their menu-items.
fixed the regular navigation for the 3 use-cases: (1) forward navigation, (2) close-button in application and (3) browser back-button.

Unfortunately, one thing remains... when a page is refreshed using the browser, the "previous-page" administration becomes overwritten with a reference to the same page, which is actually the truth. no escape possible...

erwindon commented:
@bbinet
I think I've addresses all the issues we found.
can you give it another try?

erwindon commented:
@bbinet can you please let me know whether you can try this again?

erwindon commented:
@dawidmalina you previously helped with the graphical design of SaltGUI. I was now finally able to make it into a Single Page Application. Can I invite you to try it out from this branch and let me know what you think?

bbinet commented:
@erwindon Sorry for the delay, Christmas period is very busy...
I've just tested your updated singlepage pull request again, and everything seem perfect, good job!

dawidmalina commented:
@erwindon good job. I did a few basic tests - same as i did in the past and all use cases are properly handled.

I've found few topics:
- pause icon is always present on the keys page
- pause icon is always present on the show beacons page
- run manual icon is (at least on my browser) rather elliptic

Not sure are those issues or not.

Hope i could help.

erwindon commented:
@dawidmalina 
> * pause icon is always present on the keys page
> * pause icon is always present on the show beacons page
Both the keys page and the beacons page get asynchronous updates from the salt event stream.
The pause button is a way to pause that mechanism.
(the button is also present on the events page for the same reason; and on the job-details page to stop collecting status details for each and every visible job during initial screen population or after scrolling new jobs into view)
does this explanation cover the observation?

> * run manual icon is (at least on my browser) rather elliptic
I'm quite sure nothing in this branch changed that. so likely it will be treated as a separate issue.
fyi: that 'icon' is actually the text `>_` with a big rounded green border. so technically there is no instruction present to make it a circle. but we all assume that is should be.
can you please raise this as a separate issue and including a screen-shot of that area?
very likely, I will then turn it into an svg
(It took me just 10 minutes to do that now in my workspace)

erwindon commented:
@dawidmalina
thx!
I'll push out a version+release with what was committed before this.
this will then be merged into the master branch for the next version/release.
